src/vmm/src/devices/virtio/queue.rs

```diff
@@ -242,13 +290,170 @@ impl Queue {
             desc_table_address: GuestAddress(0),
             avail_ring_address: GuestAddress(0),
             used_ring_address: GuestAddress(0),
+
+            desc_table_ptr: std::ptr::null(),
+            avail_ring_ptr: std::ptr::null_mut(),
+            used_ring_ptr: std::ptr::null_mut(),
+
             next_avail: Wrapping(0),
             next_used: Wrapping(0),
             uses_notif_suppression: false,
             num_added: Wrapping(0),
         }
     }
 
+    fn desc_table_size(&self) -> usize {
+        std::mem::size_of::<Descriptor>() * usize::from(self.size)
+    }
+
+    fn avail_ring_size(&self) -> usize {
+        std::mem::size_of::<u16>()
+            + std::mem::size_of::<u16>()
+            + std::mem::size_of::<u16>() * usize::from(self.size)
+            + std::mem::size_of::<u16>()
+    }
+
+    fn used_ring_size(&self) -> usize {
+        std::mem::size_of::<u16>()
+            + std::mem::size_of::<u16>()
+            + std::mem::size_of::<UsedElement>() * usize::from(self.size)
+            + std::mem::size_of::<u16>()
+    }
+
+    /// Set up pointers to the queue objects in the guest memory
+    /// and mark memory dirty for those objects
+    pub fn set_pointers<M: GuestMemory>(&mut self, mem: &M) -> Result<(), QueueError> {
```

Comment:
nit: Maybe this function would be better named `initialize` or `activate` of something?

Reply:
I guess `initialize` makes sense. I'll rename it later if no one has better ideas.